Akka.Serialization.MessagePack/Internal/IntIndexedMessagePackFormatterDict.cs

```diff
@@ -0,0 +1,80 @@
+using MessagePack.Formatters;
+
+namespace Akka.Serialization.MessagePack.Resolvers;
+
+/// <summary>
+/// A fast, Grow-only structure for holding Created formatters.
+/// Uses Int indexes per-type to avoid ConcurrentDictionary bucket costs,
+/// And avoids volatile/lock entirely on the happy path.
+/// </summary>
+/// <remarks>
+/// This can be (in some cases) less space efficient than a normal dictionary,
+/// However, for most internal uses, the cost is going to be fairly bounded,
+/// and <see cref="PolymorphicFormatterResolver"/>'s behavior results in
+/// <see cref="PolymorphicFormatterResolver.GetFormatter{T}"/> being a hot path,
+/// So in general the alloc cost is worth it. 
+/// </remarks>
+internal sealed class IntIndexedMessagePackFormatterDict
+{
+    private IMessagePackFormatter?[] _formatters;
+    private readonly object _lockObj = new();
+    public IntIndexedMessagePackFormatterDict(int initialSize)
+    {
+        _formatters = new IMessagePackFormatter[256];
```

Comment:
Not sure if a smaller size would be better here, I wanted to minimize re-allocs of the array but keep size 'fair'.

`256` means it's gonna be 1Kib or 2Kib for the array to start.